File: tensor/operator.go
package tensor

import (
	"github.com/lwch/gotorch/internal/torch"
)

func (t *Tensor) Backward() {
	torch.Backward(t.t, false)
}

func (t *Tensor) BackwardRetained() {
	torch.Backward(t.t, true)
}

func (t *Tensor) MatMul(t2 *Tensor) *Tensor {
	ptr := torch.MatMul(t.t, t2.t)
	return New(ptr)
}

func (t *Tensor) Add(t2 *Tensor) *Tensor {
	ptr := torch.Add(t.t, t2.t)
	return New(ptr)
}

func (t *Tensor) Sub(t2 *Tensor) *Tensor {
	ptr := torch.Sub(t.t, t2.t)
	return New(ptr)
}

func (t *Tensor) Mul(t2 *Tensor) *Tensor {
	ptr := torch.Mul(t.t, t2.t)
	return New(ptr)
}

func (t *Tensor) Div(t2 *Tensor) *Tensor {
	ptr := torch.Div(t.t, t2.t)
	return New(ptr)
}

func (t *Tensor) Pow(n float64) *Tensor {
	ptr := torch.Pow(t.t, n)
	return New(ptr)
}

func (t *Tensor) Sqrt() *Tensor {
	ptr := torch.Sqrt(t.t)
	return New(ptr)
}

func (t *Tensor) RSqrt() *Tensor {
	ptr := torch.RSqrt(t.t)
	return New(ptr)
}

func (t *Tensor) Log() *Tensor {
	ptr := torch.Log(t.t)
	return New(ptr)
}

func (t *Tensor) Exp() *Tensor {
	ptr := torch.Exp(t.t)
	return New(ptr)
}

func (t *Tensor) Neg() *Tensor {
	ptr := torch.Neg(t.t)
	return New(ptr)
}

func (t *Tensor) Abs() *Tensor {
	ptr := torch.Abs(t.t)
	return New(ptr)
}

func (t *Tensor) Max(dim int64, keepdim bool) *Tensor {
	ptr := torch.Max(t.t, dim, keepdim)
	return New(ptr)
}

func (t *Tensor) Min(dim int64, keepdim bool) *Tensor {
	ptr := torch.Min(t.t, dim, keepdim)
	return New(ptr)
}

func (t *Tensor) Sum(dim int64, keepdim bool) *Tensor {
	ptr := torch.Sum(t.t, dim, keepdim)
	return New(ptr)
}

func (t *Tensor) Mean(dim int64, keepdim bool) *Tensor {
	ptr := torch.Mean(t.t, dim, keepdim)
	return New(ptr)
}

func (t *Tensor) Var(dim int64, unbiased, keepdim bool) *Tensor {
	ptr := torch.Var(t.t, dim, unbiased, keepdim)
	return New(ptr)
}

func (t *Tensor) Relu() *Tensor {
	ptr := torch.Relu(t.t)
	return New(ptr)
}

func (t *Tensor) Gelu(tanh bool) *Tensor {
	ptr := torch.Gelu(t.t, tanh)
	return New(ptr)
}

func (t *Tensor) LeakyRelu(negSlope float64) *Tensor {
	ptr := torch.LeakyRelu(t.t, negSlope)
	return New(ptr)
}

func (t *Tensor) Silu() *Tensor {
	ptr := torch.Silu(t.t)
	return New(ptr)
}

func (t *Tensor) Sigmoid() *Tensor {
	ptr := torch.Sigmoid(t.t)
	return New(ptr)
}

func (t *Tensor) Tanh() *Tensor {
	ptr := torch.Tanh(t.t)
	return New(ptr)
}

func (t *Tensor) Softmax(dim int64) *Tensor {
	ptr := torch.Softmax(t.t, dim)
	return New(ptr)
}

func (t *Tensor) Softmax1(dim int64) *Tensor {
	ptr := torch.Softmax1(t.t, dim)
	return New(ptr)
}

func (t *Tensor) Dropout(p float64, train bool) *Tensor {
	ptr := torch.Dropout(t.t, p, train)
	return New(ptr)
}

func (t *Tensor) Unsqueeze(dim int64) *Tensor {
	ptr := torch.Unsqueeze(t.t, dim)
	return New(ptr)
}

func (t *Tensor) Squeeze(dim int64) *Tensor {
	ptr := torch.Squeeze(t.t, dim)
	return New(ptr)
}

func (t *Tensor) Flatten(startDim, endDim int64) *Tensor {
	ptr := torch.Flatten(t.t, startDim, endDim)
	return New(ptr)
}

func (t *Tensor) Contiguous() *Tensor {
	ptr := torch.Contiguous(t.t)
	return New(ptr)
}

func (t *Tensor) Expand(sizes ...int64) *Tensor {
	ptr := torch.Expand(t.t, sizes)
	return New(ptr)
}

File: lib/operator.cpp
#include <torch/torch.h>
#include "exception.hpp"
#include "operator.h"

void tensor_backward(char **err, tensor a, bool retain)
{
    auto_catch_void([a, retain]()
                    {
                        torch::Tensor gradient;
                        a->backward(gradient, retain); },
                    err);
}

tensor tensor_matmul(char **err, tensor a, tensor b)
{
    return auto_catch_tensor([a, b]()
                             { return new torch::Tensor(a->matmul(*b)); },
                             err);
}

tensor tensor_add(char **err, tensor a, tensor b)
{
    return auto_catch_tensor([a, b]()
                             { return new torch::Tensor(a->add(*b)); },
                             err);
}

tensor tensor_sub(char **err, tensor a, tensor b)
{
    return auto_catch_tensor([a, b]()
                             { return new torch::Tensor(a->sub(*b)); },
                             err);
}

tensor tensor_mul(char **err, tensor a, tensor b)
{
    return auto_catch_tensor([a, b]()
                             { return new torch::Tensor(a->mul(*b)); },
                             err);
}

tensor tensor_div(char **err, tensor a, tensor b)
{
    return auto_catch_tensor([a, b]()
                             { return new torch::Tensor(a->div(*b)); },
                             err);
}

tensor tensor_pow(char **err, tensor a, double n)
{
    return auto_catch_tensor([a, n]()
                             { return new torch::Tensor(a->pow(n)); },
                             err);
}

tensor tensor_sqrt(char **err, tensor a)
{
    return auto_catch_tensor([a]()
                             { return new torch::Tensor(a->sqrt()); },
                             err);
}

tensor tensor_rsqrt(char **err, tensor a)
{
    return auto_catch_tensor([a]()
                             { return new torch::Tensor(a->rsqrt()); },
                             err);
}

tensor tensor_log(char **err, tensor a)
{
    return auto_catch_tensor([a]()
                             { return new torch::Tensor(a->log()); },
                             err);
}

tensor tensor_exp(char **err, tensor a)
{
    return auto_catch_tensor([a]()
                             { return new torch::Tensor(a->exp()); },
                             err);
}

tensor tensor_neg(char **err, tensor a)
{
    return auto_catch_tensor([a]()
                             { return new torch::Tensor(a->neg()); },
                             err);
}

tensor tensor_abs(char **err, tensor a)
{
    return auto_catch_tensor([a]()
                             { return new torch::Tensor(a->abs()); },
                             err);
}

tensor tensor_max(char **err, tensor a, int64_t dim, bool keepdim)
{
    return auto_catch_tensor([a, dim, keepdim]()
                             { return new torch::Tensor(std::get<0>(a->max(dim, keepdim))); },
                             err);
}

tensor tensor_min(char **err, tensor a, int64_t dim, bool keepdim)
{
    return auto_catch_tensor([a, dim, keepdim]()
                             { return new torch::Tensor(std::get<0>(a->min(dim, keepdim))); },
                             err);
}

tensor tensor_sum(char **err, tensor a, int64_t dim, bool keepdim)
{
    return auto_catch_tensor([a, dim, keepdim]()
                             { return new torch::Tensor(a->sum(dim, keepdim)); },
                             err);
}

tensor tensor_mean(char **err, tensor a, int64_t dim, bool keepdim)
{
    return auto_catch_tensor([a, dim, keepdim]()
                             { return new torch::Tensor(a->mean(dim, keepdim)); },
                             err);
}

tensor tensor_var(char **err, tensor a, int64_t dim, bool unbiased, bool keepdim)
{
    return auto_catch_tensor([a, dim, unbiased, keepdim]()
                             { return new torch::Tensor(a->var(dim, unbiased, keepdim)); },
                             err);
}

tensor tensor_relu(char **err, tensor a)
{
    return auto_catch_tensor([a]()
                             { return new torch::Tensor(a->relu()); },
                             err);
}

tensor tensor_gelu(char **err, tensor a, bool tanh)
{
    return auto_catch_tensor([a, tanh]()
                             {
                                if (tanh)
                                {
                                    return new torch::Tensor(torch::gelu(*a, "tanh"));
                                }
                                return new torch::Tensor(torch::gelu(*a)); },
                             err);
}

tensor tensor_sigmoid(char **err, tensor a)
{
    return auto_catch_tensor([a]()
                             { return new torch::Tensor(a->sigmoid()); },
                             err);
}

tensor tensor_tanh(char **err, tensor a)
{
    return auto_catch_tensor([a]()
                             { return new torch::Tensor(a->tanh()); },
                             err);
}

tensor tensor_softmax(char **err, tensor a, int64_t dim)
{
    return auto_catch_tensor([a, dim]()
                             { return new torch::Tensor(a->softmax(dim)); },
                             err);
}

tensor tensor_softmax1(char **err, tensor a, int64_t dim)
{
    return auto_catch_tensor([a, dim]()
                             {
                                torch::Tensor x = *a - std::get<0>(a->max(dim, true));
                                torch::Tensor exp_x = torch::exp(x);
                                return new torch::Tensor(exp_x / (1 + exp_x.sum(dim,true))); },
                             err);
}

tensor tensor_dropout(char **err, tensor a, double p, bool train)
{
    return auto_catch_tensor([a, p, train]()
                             { return new torch::Tensor(torch::dropout(*a, p, train)); },
                             err);
}

tensor tensor_unsqueeze(char **err, tensor a, int64_t dim)
{
    return auto_catch_tensor([a, dim]()
                             { return new torch::Tensor(a->unsqueeze(dim)); },
                             err);
}

tensor tensor_squeeze(char **err, tensor a, int64_t dim)
{
    return auto_catch_tensor([a, dim]()
                             { return new torch::Tensor(a->squeeze(dim)); },
                             err);
}

tensor tensor_contiguous(char **err, tensor a)
{
    return auto_catch_tensor([a]()
                             { return new torch::Tensor(a->contiguous()); },
                             err);
}

tensor tensor_expand(char **err, tensor a, int64_t *sizes, size_t len)
{
    return auto_catch_tensor([a, sizes, len]()
                             { return new torch::Tensor(a->expand(torch::IntArrayRef(sizes, len))); },
                             err);
}